### PR TITLE
include virtual packages in hash contents

### LINKED ISSUE #4861: Include virtual packages in hash contents
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

Quoting from https://github.com/conda/conda-build/pull/4606:

> Noarch packages that use virtual packages have the virtual packages added to the hash contents of the package. This facilitates the building of noarch packages multiple times for different platforms with platform specific dependencies.
> 
> If we build `noarch` packages with diverging run dependencies for different platforms, the hash in the build string stayed the same prior to this PR, i.e., the built variants got the exact same build strings and thus package file names didn't change. With this PR we can build different variants for, e.g., `__linux`/`__osx`/`__win`, and get distinguished, non-clashing package file names.

### Why is this needed?

To move https://github.com/conda/conda-build/pull/4606 & https://github.com/conda-forge/conda-forge.github.io/pull/1839 & and much work based on that in conda-forge forward.

### What should happen?

Review & merge https://github.com/conda/conda-build/pull/4606

### Additional Context

More noarch packages = less unnecessary builds = less unnecessary downloads = less resources waste 🌍 = savings 💸

### PR DESCRIPTION
**Pull request by isuruf**
### Description

Noarch packages that use virtual packages have the virtual packages added to the hash contents of the package.
This facilitates the building of noarch packages multiple times for different platforms with platform specific dependencies.

If we build `noarch` packages with diverging run dependencies for different platforms, the hash in the build string stayed the same prior to this PR, i.e., the built variants got the exact same build strings and thus package file names didn't change.
With this PR we can build different variants for, e.g., `__linux`/`__osx`/`__win`, and get distinguished, non-clashing package file names.

Resolves #4861.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

**Comment by jaimergp** [comment]
For context, we need this as part of the work being done at https://github.com/conda-forge/conda-forge.github.io/issues/1840, following approach documented at https://github.com/conda-forge/conda-forge.github.io/pull/1839

**Comment by jakirkham** [comment]
cc @jezdez

File: tests/test_api_render.py
Comment (mbargull) [inline tests/test_api_render.py:186]:
```suggestion
    build_ids = {}
    for subdir in ['linux-64', 'linux-aarch64', 'linux-ppc64le', 'osx-64', 'win-64']:
        m = api.render(recipe_path, config=testing_config, subdir=subdir)[0][0]
        build_ids[subdir] = m.build_id()

    assert len(set(build_ids.values())) == 4
    assert build_ids['linux-64'] == build_ids['linux-aarch64']
```

File: tests/test-recipes/metadata/_noarch_with_no_platform_deps/meta.yaml
Comment (mbargull) [inline tests/test-recipes/metadata/_noarch_with_no_platform_deps/meta.yaml:2]:
```suggestion
  name: test-noarch-with-no-platform-deps
```

**Review by mbargull — DISMISSED** [review]
LGTM; just small suggestions for extending the test a little bit.

File: tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml
Comment (isuruf) [reply tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml:18]: I've gone with `__linux   # [linux]`. I don't think it's a good idea to use `ppc64le` as the selector here as the selector and the virtual package differs.

**Comment by jezdez** [comment]
I've just updated the PR description with what had been written in the news file contained in this PR and I don't understand why this wasn't done initially when this was opened.

It would help the review process, especially for the people relatively new to the code base like myself, compared to those that have plenty of experience working with it every day.

To be clear, it adds stop energy to the review process if there is:

- no issue related to the proposed change or
- the pull request description not explaining why the change is needed or
- the linked information being more pull requests elsewhere (["Packages that can be noarch"](https://github.com/conda-forge/conda-forge.github.io/issues/1840) and "[Document OS-specific approaches for noarch packages](https://github.com/conda-forge/conda-forge.github.io/pull/1839)")

If this feature request is important for conda-forge, it must be filed as an issue in this repo, to be referenced during the code review, not the least for future code archeologists and consensus finding.

I understand that filing issues has probably been dismissed as not helpful in the past, for various reasons (read: bad maintenance patterns), but I'd appreciate your help in making it easier for me and my colleagues to review your suggested changes. That would seriously help to unblock you as well I think.

Thank you for your contributions -- as always -- @isuruf!

File: tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml
Comment (mbargull) [reply tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml:18]: Yes, semantically the recipe doesn't makes sense. However, I changed it that way to let the test cover more cases, i.e., without it, we don't test whether the hash changes or not depending on if we have a virtual package dep on the same platform (OS) or not.

File: conda_build/metadata.py
Comment (jezdez) [inline conda_build/metadata.py:1568]: Why do you split the pkg here?

Comment (mbargull) [reply conda_build/metadata.py:1568]: Would this help:
```suggestion
        # add name -> match spec mapping for virtual packages
        hash_contents.update({pkg.split(" ")[0]: pkg for pkg in virtual_pkgs})
```

Comment (jaimergp) [reply conda_build/metadata.py:1568]: For context, the conda-build specs are `str` objects with up to three fields: 

```yaml
requirements:
  run:
   - pip
   - python 3.9
   - python_abi * *_cpython
```

**Comment by jakirkham** [comment]
What are the next steps here?

File: tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml
Comment (jakirkham) [reply tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml:18]: Do we have virtual architecture packages?

Comment (jezdez) [reply tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml:18]: Does the [archspec virtual package](https://github.com/conda/conda/blob/fc3371d7fe6b50df9a260a8d50ab13dc1d12ff63/conda/plugins/virtual_packages/archspec.py) count?

Comment (jakirkham) [reply tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml:18]: Interesting. How would this be used? Maybe something like this?

```suggestion
    - __osx      # [osx]
    - __archspec * ppc64le  # [ppc64le]
```

**Comment by jaimergp** [comment]
One of the tests added in this PR is failing with a `setattr` problem in the config. Maybe a bad key somewhere?

**Comment by jakirkham** [comment]
This seems to be lingering a bit. How can we help move this forward?

**Comment by isuruf** [comment]
I don't have the time to work on this PR. Anyone wants to take over?

**Comment by jakirkham** [comment]
No worries. Should add that question wasn't directed at you specifically, but to this group as a whole. It's not obvious what still needs to be done to merge. For example there are outdated review comments above. Another review might clarify what's still needed.

**Comment by beeankha** [comment]
Hi @isuruf and @jakirkham ! Several conda-core dev team members are out of office currently but I'll bring this up early next week to see if we can get this moving (or at least get a timeline/update for you). Apologies for the delay!

**Comment by jakirkham** [comment]
Thanks Bianca! 🙏

No worries. Things slip through the cracks sometimes.

Just want to make sure that we pick this one back up if possible.

**Comment by isuruf** [comment]
I think @jezdez's comment still applies

> If this feature request is important for conda-forge, it must be filed as an issue in this repo, to be referenced during the code review, not the least for future code archeologists and consensus finding.

I understand that filing issues has probably been dismissed as not helpful in the past, for various reasons (read: bad maintenance patterns), but I'd appreciate your help in making it easier for me and my colleagues to review your suggested changes. That would seriously help to unblock you as well I think.

So, someone should open an issue with a description. @mbargull added a good description for the top post here. We can copy that and add more context if needed.

**Comment by h-vetinari** [comment]
> So, someone should open an issue with a description.

Done: https://github.com/conda/conda-build/issues/4861

**Review by jezdez — APPROVED** [review]
Thanks for elaborating in the ticket and providing the additional context, everyone!

File: tests/test_api_render.py
Comment (jaimergp) [inline tests/test_api_render.py:190]: [`subdir` is a read-only property](https://github.com/conda/conda-build/blob/313fb69261135db7db0ccc7a6fa26e6d3fb426ba/conda_build/config.py#L417) so it can't be set on the `Config` object. We can set `platform` and `arch` though, so that's the fix for the failing test.

```suggestion
    for subdir_ in ["linux-64", "linux-aarch64", "linux-ppc64le", "osx-64", "win-64"]:
        platform, arch = subdir_.split("-")
        m = api.render(recipe_path, config=testing_config, platform=platform, arch=arch)[0][0]
        build_ids[subdir_] = m.build_id()
```

Comment (jaimergp) [reply tests/test_api_render.py:190]: (I think 😬)

**Comment by beeankha** [comment]
pre-commit.ci autofix

**Review by beeankha — DISMISSED** [review]

File: tests/test_api_render.py
Comment (jaimergp) [inline tests/test_api_render.py:197]:
```suggestion
    # one hash for each platform, plus one for the archspec selector
    assert len(set(build_ids.values())) == 4
    assert build_ids["linux-64"] == build_ids["linux-aarch64"]
    assert build_ids["linux-64"] != build_ids["linux-ppc64le"]  # not the same due to archspec
```

@beeankha Can we commit [this suggestion](https://github.com/conda/conda-build/pull/4606/files#r1060469957) together with this one? Thanks 🙏

File: tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml
Comment (jaimergp) [reply tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml:18]: Patched the test to cover for John's comment in [this other suggestion](https://github.com/conda/conda-build/pull/4606#discussion_r1185211731). Would that work @mbargull?

Comment (isuruf) [reply tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml:18]: That's only true when archspec is not installed and is a fallback. Otherwise you get values like power9, power10 etc.

Comment (jaimergp) [reply tests/test-recipes/metadata/_noarch_with_platform_deps/meta.yaml:18]: Hm, it passes with or without `archspec` in my dev env 🤷

**Review by beeankha — APPROVED** [review]

**Comment by jezdez** [comment]
@mbargull Any chance to re-review this?

**Review by jezdez — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks all! 🙏

**Pull request merged**
